client/db/src/lib.rs

```diff
@@ -1668,6 +1669,52 @@ impl<Block: BlockT> sc_client_api::backend::Backend<Block> for Backend<Block> {
 		Ok(())
 	}
 
+	fn append_justification(
+		&self,
+		block: BlockId<Block>,
+		justification: Justification,
+	) -> ClientResult<()> {
+		let mut transaction: Transaction<DbHash> = Transaction::new();
+		let hash = self.blockchain.expect_block_hash_from_id(&block)?;
+		let header = self.blockchain.expect_header(block)?;
+		let number = *header.number();
+
+		// Check if block is finalized first
+		let is_descendent_of = is_descendent_of(&self.blockchain, None);
```

Comment:
We can do a faster check before this as well `if number > last_finalized.number() { return error; }`.